### DISCUSSION TITLE
Better load shedding algorithm

### DISCUSSION THREAD

**Discussion by KipK** [jmcollin78/versatile_thermostat#591] [Ideas]
Sharing this idea here to improve load shedding algorithm (  Délestage in french ) 

Instead of cutting all the vtherms when power max is reached, it would be nice if the power management starts to cascading put in safety mode only some vtherms according to available power , starting by the one that has the lowest difference with setpoint. 

This way, if power shedding is needed for a long time ( imagine the EV charger use case ) , it will change dynamically the vtherms that needs to be cutted off following the temperature change on each room compared to their set point. 

The power usage should be filtered with a time constant to avoid reacting for short spikes. ( I can provide some good filtering algo for this use case we use on OpenEvse charger, I developed this part and the current shaper ). Most counters will only thrill after some time ( in France it's more than a minute )
This constant should be settable in seconds. 

Also it should triggers an  event if all the vtherms are already in safety mode or the sum of heaters power capacity won't be sufficient for the current power overshoot , then we can trigger an automation on our side to shutdown something else from HA.
And another event when this emergency mode goes back to normal.

**Comment by jmcollin78** [comment]
Hello @KipK,

Some small remarks:

> Instead of cutting all the vtherms when power max is reached,

The current algo don't cut all VTherm. The algo just check just before turning on if the power of the VTherm is available. If not, it delays the start (to the next cycle). The main point is that, all VTherm are independant, there is no "upper" intelligence above all VTherm to control them. The central mode is not and cannot easily be this intelligence (it is not a mandatory VTherm).

>  put in safety mode only

Safety mode is for VTherm without temperature sensor. Not for shedding. Shedding put the VTherm into `overpowering` state.

What you are describing here is an external component of Versatile Thermostat which can operate on Vtherm but also on other devices (maybe with turning off priority). It has been discussed before (search into discussion). This is a great way to solve this issue and this cannot be in Versatile Thermostat: in VTherm all device are totally independant of each other. It would be a mess to "cascade decision on other Vtherm".

I hope this helps for your conception.

**Reply by KipK** [reply]
#The current algo don't cut all VTherm. The algo just check just before turning on if the power of the VTherm is available. If not, it delays the start (to the next cycle). The main point is that, all VTherm are independant, there is no "upper" intelligence above all VTherm to control them. The central mode is not and cannot easily be this intelligence (it is not a mandatory VTherm).

I get it, when I've tested this all the vtherms semt to be cutted of at the same time. 

I was talking about the overpowering state, just missed the correct name. 

#What you are describing here is an external component of Versatile Thermostat which can operate on Vtherm but also on other devices

That's also what I think,  something upper level. Perhaps this could be my first dev on HA. :)

Thanks for your input.

**Reply by jmcollin78** [reply]
> Perhaps this could be my first dev on HA. :)

This is very good way to start ! I will be pleased to help if needed (in French this time and in the hacf.fr discussions)
